### ISSUE TITLE
Reconsider policy around `cargo test` without `--all-features`

### ISSUE THREAD

**Issue by sffc** [unicode-org/icu4x#3423]
Currently, several crates fail to pass tests when `cargo test` is run. They all pass with `cargo test --all-features`.

Usually when this happens it is a docstest that requires a non-default feature such as `experimental`.

There are a few ways to feature-gate docs tests. None of them are great. The least bad is likely putting some variation of the following boilerplate in the docs tests:

```
/// # #[cfg(feature = "experimental")] {
/// // test goes here
/// # } // cfg
```

The downside of making `cargo test` work across the board is:

1. Tests that fail need the above boilerplate
2. We need to enforce it in CI, likely slowing down CI

Is the improved developer experience of `cargo test` working in all crates worth the code and CI complexity?

**Comment by sffc** [comment]
Discuss with:

- @Manishearth 
- @zbraniecki 
- @sffc 

Optional:

- @robertbastian 
- @echeran

**Comment by robertbastian** [comment]
> Is the improved developer experience of cargo test working in all crates worth the code and CI complexity?

I'm not sure this would actually improve developer experience. Yes, `cargo test` wouldn't fail anymore, but only because potentially broken tests aren't actually run. Code might be broken and CI would fail later, which introduces latency.

**Comment by Manishearth** [comment]
A thing we could do is that for crates that are known to not pass tests under `--all-features`; we could introduce a library level compile error with good diagnostics (via [`compile_error!()`](https://doc.rust-lang.org/stable/std/macro.compile_error.html) when all features are not specified in test mode. I'm not sure if this would affect `test --doc` / `test --test` / `test --example`, but it would definitely show up early for anything that is running `test --lib`. 


However ultimately I don't think we need to "protect" people from the mistake of not running the full testsuite: that's what CI is for. It's latency, but not that huge a deal.

I am not strongly in favor of this, but I am not opposed to trying and keeping this clean. I would suggest having a _separate_ cronjob CI job that runs test-all-features, as well as plain `cargo test`. Or maybe just the latter, we could focus on making sure just `cargo test` works, I find it hard to see people running anything other than `cargo test` or `cargo test --all-features` (and _maybe_ `cargo test --no-default-features`). It shouldn't be the same CI run as the nightly one because I consider that an important early warning system for upstream work and we should not make it more likely for failures to be due to other causes.

I think there's a bunch of stuff like this that is not critical but nice to have that we can start enforcing in a separate cronjob CI run.

**Comment by sffc** [comment]
A risk I see is that docs tests with the boilerplate proposed in the OP will pass silently even if they aren't actually being run. This could lead to confusion where the very same test passes with some feature configurations and fails on others (runtime failure, not compile failure). It can also lead to very real situations where there is a typo in the boilerplate, a feature goes away/changes name, etc., such that these docs tests end up never actually running.

**Comment by zbraniecki** [comment]
Alternatively, we can switch those tests to be not to be run, add unit test at the bottom, and keep the doc non-runnable just for the doc sake.

My main DX focus is to move away from a scenario where someone does `cd ./components/x; cargo test` and gets failure just to later learn that "it works as intended".

**Comment by robertbastian** [comment]
> My main DX focus is to move away from a scenario where someone does cd ./components/x; cargo test and gets failure just to later learn that "it works as intended".

That someone is probably running `cargo test` after having made changes and wants to run all the tests, so I don't think hiding tests behind features is a good idea. `compile_error` is probably the best way to improve DX.

**Comment by zbraniecki** [comment]
> That someone is probably running cargo test after having made changes and wants to run all the tests

Not necessarily. I'm not sure how representative are my habits, but I find myself often cloning a rust repo, going inside and doing `cargo test` before I do anything else to verify that the crate compiles *without* my changes before I start any hacking.

**Comment by robertbastian** [comment]
In that case you probably want to run all tests though? I'm just saying that feature-gating tests can be a massive footgun.

**Comment by robertbastian** [comment]
https://github.com/rust-lang/cargo/issues/2911#issuecomment-415998960

**Comment by zbraniecki** [comment]
> In that case you probably want to run all tests though?

No, I want to do `cargo check; cargo test;` I don't spend time learning what features are there nor do I enable `--all-features`.

> I'm just saying that feature-gating tests can be a massive footgun.

I'm not advocating for that.

My proposal is to doctests that require extra features, add them as unit tests behind feature flag, and keep the doc as a doc example only (no run).

**Comment by robertbastian** [comment]
> My proposal is to doctests that require extra features, add them as unit tests behind feature flag, and keep the doc as a doc example only (no run).

This would probably be `no_compile` in most cases, in which case rustdoc shows a "this example does not compile warning".

Also if we do that then `cargo test` will not run all tests (it won't run those behind feature flags). I think it's more important to not silently ignore tests (they won't even show as `ignored`) than to make `cargo test` work without `--all-features`.

**Comment by sffc** [comment]
- @robertbastian - Now that we have compiled data as a default feature, we can run most of our tests with default features. So this mostly applies to experimental, serde/buffer, and lstm. I agree with Rust being really bad at telling you when features don't exist any more, but it's a smaller problem that it was a few weeks ago.
- @Manishearth - I'd be happy verifying only `cargo test` and `cargo test --all-features`. The risk of renaming features is a real one. We could propose rustdoc to support `feature = "a"` in the language specification. This would pave a way for Clippy to warn for this.
- @sffc - How could we enforce this on CI?
- @robertbastian - We can use the cargo workspaces command. It would also catch other problems with global feature resolution.
- @sffc - What are the steps to get this added to doc and clippy?
- @Manishearth - Rustdoc would need an RFC, but I can mentor someone. Clippy should be easier.
- @robertbastian - I'm worried that this will cause more PRs to fail on CI for issues that weren't caught locally.
- @sffc - We could also `cargo check --all-targets` instead of `cargo test`
- @Manishearth - But we need to run docstests
- @robertbastian - There's `cargo test --doc --no-run` ... but it looks like that doesn't work
- @Manishearth - The full coverage for this is `cargo check --all-targets && cargo test --doc`
- @sffc - So do we want this?
- @robertbastian - I'm slightly opposed because it makes local development slightly different from CI
- @Manishearth - I'm also slightly opposed
- @sffc - I feel like this should be a best-effort, but there's no way to make it a best-effort without enforcing it somehow on CI. Could we make this an explicit list of crates?
- @Manishearth - I like that
- @robertbastian - Can @zbraniecki reiterate what we gain by doing this?
- @zbraniecki - My hypothesis is that I'm not the only person who does a sanity check who clones a crate and runs `cargo test` to make sure it works.
- @robertbastian - So this is more related to onboarding ease of use
- @Manishearth - From a dev-dependency point of view, is it possible to enable your own features?
- @sffc - There are instances in crates like icu_datetime where we have code paths that are only reachable with a subset of features
- @robertbastian - If you're making a change to the crate, at some point you need to figure out you need to run `--all-features`.
- @sffc - I do feel that it is valuable to reduce the barrier for entry for new developers. It is a real problem that `cargo test` does not work; it is a barrier in the way of contributors.
- @robertbastian - We could make a cfg'd compile_error that triggers on the features missing.
- @Manishearth - But there will be a whole bunch of other errors too so the important one may be hard to find
- @robertbastian - I'd rather have PRs "just work" when uploaded.
- @sffc - We already get errors a lot on tidy, clippy, fmt, ...
- @robertbastian - Yeah but `--all-features` could hide real problems that could be real work to fix.
- @zbraniecki - Hmm, can we alias `cargo test` to `cargo test --all-features` ?
- @robertbastian - It's not possible (cites a docs page)
- @Manishearth - We could write a script that make sure Cargo.toml features are consistent with feature-gated docs tests
- @sffc/@Manishearth - I think it will be uncommon for things to fail on `--all-features` that don't fail without it in general
- @zbraniecki - It's fine for it to fail in CI if it has a feature disabled
- @robertbastian - It just seems like this doesn't have good ROI
- @sffc - We only need to feature-gate individual docs tests if they are not attached to a type that is not feature gated

Proposals:

1. Make a CI that runs a custom command within an explicit list of crates
  - Normally it will run `cargo check --all-targets && cargo test --doc` or in some cases `cargo test-all-features` or other arbitrary commands
  - Only do this for crates where it is easy to do
  - Try to avoid feature-gating the docs tests; if we do it, it would be good to make a tool that enforces the feature gating of docs tests
2. Add a cfg'd compile_error that prints out a nice error message when running `cargo test` without `--all-features`
  - But the error is hard to find. See also https://github.com/rust-lang/rust/issues/68838

- @robertbastian - I am opposed to adding CI for this because it's work to maintain and we're good enough with compiled data being a default feature
- @Manishearth - We should leave the door open to discussing this again if it ends up being an issue

New proposal:

- Add the CI as proposed above in 1 on an optional, opt-in basis for individual crates
- We will revisit this CI if it ends up hindering our workflows

LGTM: @Manishearth @sffc @skius @robertbastian
